### PR TITLE
Update filter for Kotlin coroutines that restore state

### PR DESCRIPTION
**Pull request by Godin**

### CODE REVIEW

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java [hunk 1]
```diff
@@ -72,32 +73,43 @@ private void match(final MethodNode methodNode,
...
 
-			for (AbstractInsnNode i = methodNode.instructions
-					.getFirst(); i != null; i = i.getNext()) {
+			int suspensionPoint = 1;
+			for (AbstractInsnNode i = cursor; i != null
+					&& suspensionPoint < s.labels.size(); i = i.getNext()) {
 				cursor = i;
 				nextIsVar(Opcodes.ALOAD, "COROUTINE_SUSPENDED");
 				nextIs(Opcodes.IF_ACMPNE);
+				AbstractInsnNode continuationAfterLoadedResult = cursor;
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java]: At this point we know that cursor is `JumpInsnNode`. Cast happens later, which confused me. Also `continuationAfterLoadedResult` first points to jump insn and then later gets replaced with jump target. Both can be fixed if we replace this line with:


	if (cursor == null) {
		continue;
	}
	final AbstractInsnNode continuationAfterLoadedResult = skipNonOpcodes(((JumpInsnNode) cursor).label);

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/AbstractMatcher.java [hunk 2]
```diff
@@ -158,11 +158,21 @@ final void next() {
...
 	final void skipNonOpcodes() {
+		cursor = skipNonOpcodes(cursor);
+	}
+
+	/**
+	 * Returns first instruction from given and following it that is not
+	 * {@link AbstractInsnNode#FRAME}, {@link AbstractInsnNode#LABEL},
+	 * {@link AbstractInsnNode#LINE}.
+	 */
+	final AbstractInsnNode skipNonOpcodes(AbstractInsnNode cursor) {
```
Comment (marchof) [inline org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/AbstractMatcher.java]: As this method does not change state (unlike the others) I would prefer to declare it `static`.

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java (hunk 1 above)
Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java]: Believe me - this is exactly how first implementation looked, then I decided to make flow linear 😆

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/AbstractMatcher.java (hunk 2 above)
Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/AbstractMatcher.java]: Indeed. Fixed!

File: org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java (hunk 1 above)
Comment (Godin) [reply org.jacoco.core/src/org/jacoco/core/internal/analysis/filter/KotlinCoroutineFilter.java]: Fixed.

**Pull request merged**